### ISSUE TITLE
Pluto 23v3.1 QA issues

### ISSUE THREAD

**Issue by fvankrieken** [NYCPlanning/data-engineering#405]
@jackrosacker flagged on teams earlier this week

@AmandaDoyle fyi

> Flagging an issue we've found in our MapPLUTO QA. Looks like some tax lot changes have been incorporated into the most recent 23v3.1 minor build. Briefly, it looks like the 23v3.1 build has 6 lots more than the 23v3 build, and further that there is a mismatch of ~60 BBLs when joining on that column. From an initial look, the mismatches look mostly to do with lot splits/merges and condo creation.

> Version | Source | Record Count
> -- | -- | --
> 23v3.1 | source | 857,042
> 23v3 | source | 857,036

23v3 being pulled from published version 23v3, 23v3.1 from draft `fvk-pluto-versioning`. These record counts are using the output `mappluto_unclipped_gdb.gdb.zip`

A couple specific bbls - from @jackrosacker querying data loaded to the gis sde, the following bbls were present in 23v3 but not in 23v3.1

1001240008
1007380008
1013460012
1014290003
1014290044
1015700017
1021520083
2025620049
2027240032
2029460047
2030920003
2031430282
2044270013
2044270014
2044270015
2044270016
2059060342
3001190055
3002787502
3004350047
3009340069
3010400013
3011120012
3018400001
3023660032
3025200001
3036870020
3036970033
3048140026
3049940013
3056210005
3056820045
3063250030
3064110022
3067670090
4004300008
4004300010
4004300011
4004300012
4009100001
4010780051
4050667502
4050667505
4076217501
5010080049
5014957501
5024017501
5024507504
5030030025
5030247501
5031737504
5038320021
5065800056
5065800058
5070557502
5079390056
5080400021

A specific case, there's a polygon that seems to be consistent yet change bbl from 3002787502 to 3002787501

**Comment by fvankrieken** [comment]
I've verified that source data versions are consistent for all that should be - only zoning data sources are updated from 23v3 to 23v3.1

**Comment by fvankrieken** [comment]
Data available to be investigated right now
- build outputs for both 23v3 and 23v3.1 in DO
  - clipped and unclipped shapefiles
  - clipped and unclipped gdbs
  - pluto csv
- intermediate tables for build of 23v3.1 in edm-data (schema fvk-pluto-versioning@db-pluto)
- pluto 23v3 in edm-data (dcp_pluto.23v3@defaultdb). I BELIEVE this is equivalent to the "pluto" table at the end of a build but need to confirm

**Comment by fvankrieken** [comment]
I've done a bit of digging which I'll try to summarize in a bit, but as a bit of a first step am going to rebuild pluto with identical inputs to 23v3. We'll first need to compare outputs to 23v3, identify anything that is truly nondeterministic. Then we can start comparing intermediate tables to 23v3.1 and try to figure out where issues are arising from, whether they are from semi-nondeterministic code (akin to condo proxcode issue) or "intended" behavior - some sort of filtering at some step that new zoning data rightfully changes for a certain lot, etc

**Comment by fvankrieken** [comment]
Resolved 3002787502 morphing into 3002787501. These are billing/primebbls which share a base_billing_bbl (in dtm_condo). dtm only has one polygon for the basebbl and nothing smaller that could correspond to 7501 and 7502. @AmandaDoyle thinks this is likely a dtm issue, that hopefully it will "catch up" with reality at some point soon.

Nothing should be done to code, since multiple base bbls per billing is not valid. However, it might be worth adding code to at least make this deterministic (grab smallest primebbl associated with basebbl). 

Logic around this is in `dtmmergepolygons.sql`

**Comment by fvankrieken** [comment]
Simpler issue than I thought, at least in terms of why the source data seems different

<img width="633" alt="image" src="https://github.com/NYCPlanning/data-engineering/assets/9454672/0fda114a-e6ec-4065-853b-96a1d75bff01">

If it looks like a lot change, it's probably a lot change.

We recently discussed that when pushing draft outputs, we should be cleaning (deleting) any existing files in s3 prior to pushing. This was an insidious case of a [silent failure](https://github.com/NYCPlanning/data-engineering/actions/runs/6552814120/job/17796915237#step:14:1272) in the October build on @sf-dcp 's branch, where the gdb export failed. `set e` either isn't working as we intended within the function call (we need to maybe explicitly capture the gdal error and then exit), or gdal itself failed silently. But outputs were made MINUS the gdbs, which were then pushed to s3. Old outputs overwritten except for the gdbs, which remained
